### PR TITLE
Provide parallel_one_host_only via workers config file

### PR DESCRIPTION
**Pull request by b10n1k**
https://progress.opensuse.org/issues/158146

### CODE REVIEW

File: docs/WritingTests.asciidoc
Comment (okurz) [inline docs/WritingTests.asciidoc:676]:
```suggestion
enable this. This setting can be applied as a test variable during the time
of scheduling as well as in the worker configuration file `workers.ini`.
```

File: etc/openqa/workers.ini
Comment (okurz) [inline etc/openqa/workers.ini:8]: no

File: etc/openqa/workers.ini
Comment (okurz) [inline etc/openqa/workers.ini:11]:
```suggestion
# Enable to restrict any multi-machine jobs taken by this worker host to only run here.
# Disabled by default.
#PARALLEL_ONE_HOST_ONLY = 0
```

File: etc/openqa/workers.ini
Comment (okurz) [inline etc/openqa/workers.ini:97]: no

**Review by okurz — CHANGES_REQUESTED** [review]

File: etc/openqa/workers.ini
Comment (Martchus) [reply etc/openqa/workers.ini:11]:
```suggestion
# Enable to restrict any multi-machine jobs taken by this worker host to only run there.
# Disabled by default.
#PARALLEL_ONE_HOST_ONLY = 0
```

File: lib/OpenQA/WebAPI/Controller/API/V1/Worker.pm
Comment (Martchus) [inline lib/OpenQA/WebAPI/Controller/API/V1/Worker.pm:187]:
```suggestion

```

File: lib/OpenQA/WebAPI/Controller/API/V1/Worker.pm
Comment (Martchus) [inline lib/OpenQA/WebAPI/Controller/API/V1/Worker.pm:179]:
```suggestion
      for qw(cpu_modelname cpu_opmode cpu_flags isotovideo_interface_version job_id websocket_api_version parallel_one_host_only);
```

**Review by Martchus — CHANGES_REQUESTED** [review]

File: etc/openqa/workers.ini
Comment (b10n1k) [reply etc/openqa/workers.ini:97]: it is dumped PR. those are not meant to stay obviously

File: t/24-worker-overall.t
Comment (Martchus) [inline t/24-worker-overall.t:158]: This needs to be in alphabetical order. But to not repeat the whole previous subtest this test should probably just check `is $capabilities->{parallel_one_host_only}, 1, …`.

File: t/24-worker-overall.t
Comment (Martchus) [inline t/24-worker-overall.t:145]:
```suggestion
    subtest 'capabilities read PARALLEL_ONE_HOST_ONLY setting' => sub {
```

File: t/24-worker-overall.t
Comment (Martchus) [inline t/24-worker-overall.t:147]: You probably need to set the `PARALLEL_ONE_HOST_ONLY` setting in `$global_settings` before invoking this function.

**Review by Martchus — CHANGES_REQUESTED** [review]

File: t/24-worker-overall.t
Comment (b10n1k) [reply t/24-worker-overall.t:147]: But if I am not wrong it should read it from https://github.com/os-autoinst/openQA/pull/5695/files#diff-90a20a69fe7c107578278a313173bef8c87bade502ecec03004c6c02eadebeeaR13

Comment (Martchus) [reply t/24-worker-overall.t:147]: No, it won't. Only if the instance matches and you edit the settings file that is actually relevant for this test. So far your edits don't change a relevant combination.

Note that instead of changing those ini files you can also just modify `$global_settings`. At least I think so judging by how the subsequent subtest does it.

Comment (b10n1k) [reply t/24-worker-overall.t:147]: worked. I had a mistake in the code

File: t/24-worker-overall.t
Comment (b10n1k) [reply t/24-worker-overall.t:145]: Done

File: docs/WritingTests.asciidoc
Comment (b10n1k) [reply docs/WritingTests.asciidoc:676]: Done

File: lib/OpenQA/WebAPI/Controller/API/V1/Worker.pm
Comment (b10n1k) [reply lib/OpenQA/WebAPI/Controller/API/V1/Worker.pm:187]: Done

File: docs/WritingTests.asciidoc
Comment (Martchus) [reply docs/WritingTests.asciidoc:676]: I don't see the suggestion by @okurz applied (but would also prefer his phrasing).

File: etc/openqa/workers.ini
Comment (Martchus) [reply etc/openqa/workers.ini:11]: None of our suggestions has been applied.

File: etc/openqa/workers.ini
Comment (Martchus) [reply etc/openqa/workers.ini:97]: I suggest you use https://open.qa/docs/#_customize_configuration_directory for tests like that. If you use my helper setup: It makes already use of this variable so editing this file here won't do anything for you anyway.

File: lib/OpenQA/WebAPI/Controller/API/V1/Worker.pm
Comment (Martchus) [reply lib/OpenQA/WebAPI/Controller/API/V1/Worker.pm:187]: No yet, actually.

**Review by Martchus — CHANGES_REQUESTED** [review]

File: docs/WritingTests.asciidoc
Comment (b10n1k) [reply docs/WritingTests.asciidoc]: yes, apparently I forgot to push. most of them should be there now

File: t/24-worker-overall.t
Comment (Martchus) [inline t/24-worker-overall.t:151]:
```suggestion
```

File: t/24-worker-settings.t
Comment (Martchus) [inline t/24-worker-settings.t:128]:
```suggestion
```

I don't think we need that test. You haven't changed anything in the settings parser itself and we should not add such a big test for every setting we introduce. (Or are changes in the settings parser still going to be needed?)

**Review by Martchus — CHANGES_REQUESTED** [review]

File: t/data/24-worker-overall/workers.ini
Comment (okurz) [inline t/data/24-worker-overall/workers.ini:13]: I suggest to not change this fixtures file but change the settings within the (sub-)test where you need it

**Review by okurz — CHANGES_REQUESTED** [review]

File: t/data/24-worker-overall/workers.ini
Comment (b10n1k) [reply t/data/24-worker-overall/workers.ini]: Done

File: t/24-worker-settings.t
Comment (b10n1k) [reply t/24-worker-settings.t]: Done

File: t/24-worker-overall.t
Comment (b10n1k) [reply t/24-worker-overall.t]: Done

**Comment by b10n1k** [comment]
All the unnecessary changes have been reverted. I couldnt extend the tests further but the existing coverage should be satisfied for now.

File: docs/WritingTests.asciidoc
Comment (Martchus) [inline docs/WritingTests.asciidoc:676]: The whole feature hasn't been implemented yet so adding the documentation now is a bit problematic (as it would not really work just yet).

**Review by Martchus — COMMENTED** [review]
Looks good but I'd move out the two documentation changes just for now.

File: docs/WritingTests.asciidoc
Comment (okurz) [reply docs/WritingTests.asciidoc:676]: We won't merge this PR in this stage but I consider the approach to write the documentation first :)

**Review by okurz — CHANGES_REQUESTED** [review]

**Review by okurz — CHANGES_REQUESTED** [review]

File: lib/OpenQA/Schema/Result/Workers.pm
Comment (Martchus) [reply lib/OpenQA/Schema/Result/Workers.pm:84]: This change is probably not required at all.

**Review by Martchus — COMMENTED** [review]
I still think this lacks some additions to the scheduler logic. I'll come up with a test case next week to show that.

**Comment by Martchus** [comment]
I pushed a few more commits to add tests for the scheduler logic and extend it. It was not completely trivial because the evaluation of the setting now needs to be done in a much more dynamic way. I haven't done any manual tests yet.

**Review by Martchus — APPROVED** [review]
With my changes I find it good enough to be merged. I am keeping the "not-ready" label until we/I did some manual testing.

File: docs/WritingTests.asciidoc
Comment (Martchus) [reply docs/WritingTests.asciidoc:676]: With my latest changes we can include this documentation.

File: lib/OpenQA/Scheduler/WorkerSlotPicker.pm
Comment (b10n1k) [inline lib/OpenQA/Scheduler/WorkerSlotPicker.pm:65]: Seems like this line is duplicated. the ` _id_or_skip` does the same check which will return `undef` and it will skip this line anyway. Probably it will never run.

**Comment by Martchus** [comment]
I now tested this manually with a cluster of 2 jobs and all workers with `PARALLEL_ONE_HOST_ONLY=1` as we'd use it in production. None of the jobs had the setting.

* With all workers on different hosts nothing is scheduled as expected (as we need at least 2 workers with a common host name).
* As soon as I replaced one worker so that there are now two workers with a common host jobs were scheduled on those two slots as expected.
* When I replaced one more worker so that there are now two hosts with two slots each in the pool I was able to occupy all worker slots and clusters were still not scheduled across multiple hosts.

So I guess that's good enough.

File: etc/openqa/workers.ini
Comment (okurz) [inline etc/openqa/workers.ini:9]: wrap

**Review by okurz — APPROVED** [review]
One minor issue in comments, rest fine. Feel welcome to self-approve after having fixed that.

**Comment by Martchus** [comment]
Really annoying that codecov uploads occaisonally fail:

```
debug - 2024-07-02 09:24:29,772 -- Upload result --- {"result": "RequestResult(error=RequestError(code='HTTP Error 401', params={}, description='{\"detail\":\"Invalid token header. No credentials provided.\"}'), warnings=[], status_code=401, text='{\"detail\":\"Invalid token header. No credentials provided.\"}')"}
error - 2024-07-02 09:24:29,772 -- Upload failed: {"detail":"Invalid token header. No credentials provided."}
```

Since this isn't treated as an error (the CircleCI codecov step passes) I'll have to re-run all CircleCI tests.

**Comment by Martchus** [comment]
I retried yesterday and also today and Codecov is reliably broken. That's also the case on other PRs like https://github.com/os-autoinst/openQA/pull/5723.

Strangely, OBS doesn't have the same problem. I checked the upload steps of a few PRs on https://github.com/openSUSE/open-build-service/pull. Maybe we should compare our setup with the one from OBS.

For the time being I checked https://output.circle-artifacts.com/output/job/5a4e33c3-6409-401d-b492-b79ba22e4c05/artifacts/0/cover_db/coverage.html manually and it looks good. So I'm merging this PR now.

**Pull request merged**
